[extensions][py][hf] 2/n ASR model parser impl

Ankush-lastmile opened the pull request:
[extensions][py][hf] 2/n ASR model parser impl




Model Parser for the Automatic Speech Recognition task on huggingface.


Decisions made while implementing:
- manual impl to parse input attachments
- - threw exceptions on every unexpected step. Not sure if this is the direction we want to go with this.
- This diff does not implement serialize() for the model parser (will implement on diff ontop)

## Testplan

Created an mp3 file that says "hi". Used aiconfig to run asr on it.

|<img width="922" alt="Screenshot 2024-01-09 at 7 14 47 PM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/fe68751d-e20b-41d9-9da5-cc9a32859cba">

|<img width="1461" alt="Screenshot 2024-01-09 at 5 54 33 PM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/78063a3e-2b9a-4a39-80d9-ef28a7d706cf">|
| ------------- | ------------- |

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/lastmile-ai/aiconfig/pull/780).
* __->__ #780
* #779

jonathanlastmileai commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/local_inference/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,152 @@
+
+from transformers import pipeline, Pipeline
+
+if TYPE_CHECKING:
+    from aiconfig import AIConfigRuntime
```
Totally separate discussion: do we have to do this?

Ankush-lastmile replied:
from before, this was needed. Keeping it for now since we don't have automated tests running

saqadri commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/__init__.py:
```diff
@@ -1,8 +1,9 @@
 from .local_inference.text_2_image import HuggingFaceText2ImageDiffusor
 from .local_inference.text_generation import HuggingFaceTextGenerationTransformer
-from .remote_inference_client.text_generation import HuggingFaceTextGenerationClient
+# from .remote_inference_client.text_generation import HuggingFaceTextGenerationClient
```
Remove?

rossdanlm replied:
Why does this keep erroring or everyone lol?

Ankush-lastmile replied:
will remove after rebase

saqadri commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/local_inference/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,152 @@
+
+    audio_uris: list[str] = []
+
+    for i, attachment in enumerate(prompt.input.attachments):
```
Let's match the type of attachment to be the same as
 
```
class OutputDataWithStringValue(BaseModel):
    """
    This represents the output content that is storied as a string, but we use
    both the `kind` field here and the `mime_type` in ExecuteResult to convert
    the string into the output format we want.
    """

    kind: Literal["file_uri", "base64"]
    value: str
```

So instead of data being the uri itself, would be great for data to be {"kind": "file_uri", "value": "<uri>"}

We can add proper typings for attachments later, but having that type for this model parser will help us standardize on this format

rossdanlm replied:
For more context, you can see how we did it for (look up `kind=` in these file changes):
- image output: https://github.com/lastmile-ai/aiconfig/pull/608/commits/71749931e095a8cdf1e07e2432f36401f647cf7a
- tool call output: https://github.com/lastmile-ai/aiconfig/pull/611/files

The output schema is defined in https://github.com/lastmile-ai/aiconfig/blob/main/python/src/aiconfig/schema.py#L71-L74

Ankush-lastmile replied:
as per additional converstion offline, added custom input type `AttachmentInputDataWithStringValue` inside this model parser. Enforces file attachments to be of this type containing a `kind` value. 

Added todo's to cleanup and add to aiconfig sdk

Ankush-lastmile commented:
Added a couple of changes:

1. Defined `AttachmentInputDataWithStringValue`. This will need to be incorporated into the AIConfig sdk but for now we will define this inside this model parser and enforce it here. TODO: Task #

1.5. AIConfig for this will look slightly different. See updated testplan

2. Updated Model Parser to be able to customize pipeline creation. Note: for now this customizability is only possible the first time the pipeline is created in memory

3. Updated Model Parser completion params with certain subset

Ankush-lastmile commented:
rebase, and fix init.py

rholinshead commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/local_inference/automatic_speech_recognition.py:
```diff
@@ -52,14 +55,224 @@ async def serialize(
+                f"Error: {str(e)}. Please specify a kind and value for the attachment data."
+            )
+        # TODO: once previous todo gets resolved, modify this line to `.kind`` instead of .data.get("kind"). It will be a pydantic base class.
+        if not attachment.data.get("kind") == "file_uri":
```
We'll need to update the client rendering and prompt schema format to support this. And we'll need to be consistent with this type in the image-to-text parser as well

rholinshead replied:
Can we just have this be a string and migrate to the object with kind and value afterwards? My concerns here are:
- client rendering is currently adding attachments as {data: <url>, mime_type}
- image-to-text parser is currently not doing this {kind/value} stuff and that inconsistency is going to cause a lot more work on the client

Ankush-lastmile replied:
discussed offline with @rholinshead, reverted change to input data type. For ease of use we will have data as just a string and implement structured input later

Ankush-lastmile commented:
discussed offline with @rholinshead, reverted change to input data type. For ease of use we will have data as just a string and implement structured input later

updated testplan to show as well

rossdanlm commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/local_inference/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,257 @@
+        await aiconfig.callback_manager.run_callbacks(CallbackEvent("on_run_complete", __name__, {"result": prompt.outputs}))
+        return prompt.outputs
+
+    def _get_device(self) -> str:
+        if torch.cuda.is_available():
+            return "cuda"
+        # Mps backend is not supported for all asr models. Seen when spinning up a default asr pipeline which uses facebook/wav2vec2-base-960h 55bb623
+        return "cpu"
```
We should use mac if we have it. See `_get_device` from text_2_image.py

rossdanlm commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/local_inference/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,257 @@
+        return "cpu"
+
+    def get_output_text(self, response: dict[str, Any]) -> str:
+        raise NotImplementedError("get_output_text is not implemented for HuggingFaceAutomaticSpeechRecognition")
```
Why are we not implementing?

Ankush-lastmile replied:
added implementation

rossdanlm commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/local_inference/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,257 @@
+    TODO: Figure out how to find which params are supported.
+    """
+
+    supported_keys = {
+        # inputs
+        "return_timestamps",
+        "generate_kwargs",
+        "max_new_tokens",
+    }
```
There's actually way more because we're using `generate_kwargs` which takes in from the generalized text generation params. Pls add this to the task for unifying this in prompt schema, and link to task here

rossdanlm commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/local_inference/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,257 @@
+    """
+    Constructs an output from the response of the HF ASR pipeline.
+
+    Response from pipeline could contain multiple outputs and time stamps
```
Can you provide link to API so it's easier to tell what examples they can be?

rossdanlm commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/local_inference/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,257 @@
+        response = [response]
+
+    for i, result in enumerate(response):
+        # response is expected to be a dict containing the text output and timestamps if specified. Could not find docs for this.
```
rip, ok pls mention can't find docs in the method docstrings

Ankush-lastmile replied:
updated

rossdanlm commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/local_inference/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,257 @@
+        output = ExecuteResult(
+            **{
+                "output_type": "execute_result",
+                "data": result.get("text"),
```
So response is always guaranteed to have the text attribute? I think it would be safe just to have another line above saying `text_output: str = result.get("text") if "text" in result and isinstance(result, dict) else result` or whatever just to future proof

rossdanlm replied:
The problem is without knowing whta `result` type is, we can't know for sure and better to be safe just to ensure it can't break

rossdanlm commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/local_inference/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,257 @@
+            device = self._get_device()
+            if pipeline_creation_data.get("device", None) is None:
+                pipeline_creation_data["device"] = device
+            self.pipelines[model_name] = pipeline(task="automatic-speech-recognition", **pipeline_creation_data)
```
nit: no change, just annoying that they called this `automatic-speech-recognition` instead of `audio-to-text` like everything else they did smh

rossdanlm commented on extensions/HuggingFace/python/src/aiconfig_extension_hugging_face/local_inference/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,257 @@
+"""
+
+
+class HuggingFaceAutomaticSpeechRecognition(ParameterizedModelParser):
```
nit: add "Transformer" suffix at end to meet same format as the others

Ankush-lastmile replied:
updated

rossdanlm reviewed (approved):
pls link relevant comments to relevant issues so we can track in the future

Ankush-lastmile commented:
- added some comments
- renamed to `HuggingFaceAutomaticSpeechRecognitionTransformer` from `HuggingFaceAutomaticSpeechRecognitionTransformer`
- added get output text impl (copied from other model parser impls

Ankush-lastmile commented:
rebase
